### PR TITLE
drivers: dma: Add Xilinx AXI DMA driver

### PR DESCRIPTION
**Pull request by WorldofJARcraft**
The [Xilinx AXI DMA Controller](https://www.xilinx.com/products/intellectual-property/axi_dma.html) is commonly used in FPGA designs. For example, it is a part of the [1G/2.5G AXI Ethernet subsystem](https://www.xilinx.com/products/intellectual-property/axi_ethernet.html). This patch adds a driver for the Xilinx AXI DMA that supports single MM2S and S2MM channels as well as the control and status streams used by the AXI Ethernet subsystem.
I will send a pull request for MAC and MDIO drivers for the AXI Ethernet subsystem shortly.

### CODE REVIEW

File: dts/bindings/dma/xilinx,eth-dma.yaml
Comment (decsny) [inline dts/bindings/dma/xilinx,eth-dma.yaml:1]: why is a different compatible needed?

Comment (WorldofJARcraft) [reply dts/bindings/dma/xilinx,eth-dma.yaml:1]: This comes down to an idiosyncrasy in the Xilinx tools: Vitis can automatically generate a device tree from an (Vivado) FPGA project. This is useful, because MMIO addresses are assigned in the FPGA project and need to be reflected in the device tree. I have noticed that when the DMA is paired with any core *other* than the AXI Ethernet MAC core from the AXI Ethernet subsystem in the FPGA project, Vitis will generate a compatible string `xlnx,axi-dma-1.00.a`. However, if the DMA core is used with the MAC core, Vitis will generate a compatible `xlnx,eth-dma`. The main difference between using the DMA with the Ethernet MAC vs. with any other AXI-Stream module is that in the Ethernet subsystem, the DMA has additional control and status streams that are currently only used for checksum offloading. I hope that gets more clear when I send pull requests for the MAC driver next week.
Whether these additional streams are present is also encoded in the `axistream-control-connected` property, making the different compatible strings redundant. However, currently (at least until version 2023.2), Vitis will still generate two different compatible strings for the DMA.
I elected to allow both compatibles, because especially on the Zynq MPSOC, someone might use the Vitis-generated device tree for both the first-stage bootloader or u-boot (which is necessary for programming the DMA) and in Zephyr.

Comment (WorldofJARcraft) [reply dts/bindings/dma/xilinx,eth-dma.yaml:1]: I have submitted a pull request for the MAC driver: https://github.com/zephyrproject-rtos/zephyr/pull/73986

**Review by teburd — DISMISSED** [review]

File: dts/bindings/dma/xilinx,axi-dma-base.yaml
Comment (decsny) [inline dts/bindings/dma/xilinx,axi-dma-base.yaml:34]: why does it default to 32 bit? should it just be required: true?

Comment (WorldofJARcraft) [reply dts/bindings/dma/xilinx,axi-dma-base.yaml]: Good point, I have changed this accordingly.

**Comment by saizen408** [comment]
@WorldofJARcraft  Is it on purpose that `zephyr_library_sources_ifdef(CONFIG_XILINX_AXI_DMA dma_xilinx_axi_dma.c)` is not included in `zephyr/drivers/dma/CMakeLists.txt`? 

I'm fairly new to zephyr so forgive if I am missing something.

**Review by teburd — DISMISSED** [review]

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (teburd) [inline drivers/dma/dma_xilinx_axi_dma.c:769]: configuration errors shouldn't be a k_oops(), the error should be propogated up the callstack to the application which can do what it likes, perhaps asserting or panicking (k_oops())

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (teburd) [inline drivers/dma/dma_xilinx_axi_dma.c:864]: __ASSERT() instead, this is a programming error

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (teburd) [inline drivers/dma/dma_xilinx_axi_dma.c:640]: If this is unimplemented (along with resume) set the function pointers to NULL in the API struct instead, this is dealt with at an API level for you already 

e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/dma.h#L501

**Review by teburd — CHANGES_REQUESTED** [review]
Some small changes could make this driver a bit more Zephyr-like.

k_oops() is pretty rare to see in a driver, generally if its a programming error __ASSERT(), if its a runtime error propogate the error code up the call stack along with a reasonable log message to inform whats going on.

File: drivers/dma/Kconfig.xilinx_axi_dma
Comment (WorldofJARcraft) [reply drivers/dma/Kconfig.xilinx_axi_dma:27]: I changed it to 16, which is what I use on my test platform.
As the DMA is used with a variety of different peripherals, I expect users of this driver to overwrite the value with a value suitable for their specific application anyway.

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (WorldofJARcraft) [reply drivers/dma/dma_xilinx_axi_dma.c:504]: Depending on the application, this function might be called from interrupt handlers triggered by different devices, periodically using threads or timers or in any other possible context. Thus, I think locking all IRQs generally is the safest action.
I will wrap this into a configuration variable, such that users of the driver can choose to only mask the DMA interrupt on the same or both channels, respectively, if their application allows it.

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (WorldofJARcraft) [reply drivers/dma/dma_xilinx_axi_dma.c:769]: I have removed the oops() and use appropriate error messages / return values instead.

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (WorldofJARcraft) [reply drivers/dma/dma_xilinx_axi_dma.c:864]: Done

File: drivers/dma/dma_xilinx_axi_dma.c
Comment (WorldofJARcraft) [reply drivers/dma/dma_xilinx_axi_dma.c:640]: Done

**Comment by WorldofJARcraft** [comment]
I have noticed that the DMA in conjunction with the Ethernet driver (https://github.com/zephyrproject-rtos/zephyr/pull/73986) stops receiving Ethernet packets at some point when CONFIG_NET_TC_RX_COUNT is not 0. Please do not merge while I investigate this.

**Comment by WorldofJARcraft** [comment]
> I have noticed that the DMA in conjunction with the Ethernet driver (#73986) stops receiving Ethernet packets at some point when CONFIG_NET_TC_RX_COUNT is not 0. Please do not merge while I investigate this.

This is fixed; the cause was resetting the DMA multiple times.

**Review by teburd — DISMISSED** [review]

File: dts/bindings/dma/xilinx,eth-dma.yaml
Comment (decsny) [inline dts/bindings/dma/xilinx,eth-dma.yaml:5]: why can't the same string be used?

**Review by decsny — CHANGES_REQUESTED** [review]

File: dts/bindings/dma/xilinx,eth-dma.yaml
Comment (WorldofJARcraft) [reply dts/bindings/dma/xilinx,eth-dma.yaml:5]: Xilinx' Vitis tool can automatically generate device trees. This feature is often used as a starting point before further customization. The output generated by Vitis is context-dependent; for DMAs used together with the Ethernet subsystem, it generates an `eth-dma` compatible, for other uses it generates the other, more generic, compatible.
Linux actually binds different drivers for the DMA depending on the compatible: the entirety of the Ethernet subsystem including DMA has one driver, and there is an independent driver for the stand-alone DMA (e.g., useful for custom IPs connected to the DMA).

Comment (WorldofJARcraft) [reply dts/bindings/dma/xilinx,eth-dma.yaml:5]: Small correction: Upstream Linux actually does not actively *use* the `eth-dma` compatible. Instead, the Ethernet driver resolves the dma using the *axistream-connected* handle. Still, using the `eth-dma` handle prevents the generic DMA driver from claiming the device. 
References:
- Ethernet driver: https://github.com/torvalds/linux/blob/master/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
- DMA driver: https://github.com/torvalds/linux/blob/master/drivers/dma/xilinx/xilinx_dma.c
- My other PR https://github.com/zephyrproject-rtos/zephyr/pull/77732 has an exemplary device tree based on what Vitis generated automatically (notice the `eth-dma` compatible): https://github.com/zephyrproject-rtos/zephyr/blob/eda4bedfb96b2f07dd65142dc335d247e680b971/dts/riscv/openhwgroup/cv64a6.dtsi

**Comment by WorldofJARcraft** [comment]
I would like to do additional testing before this is eventually merged.
Should I convert this pull request back to a draft, or is it better to add a Do Not Merge label?

**Comment by decsny** [comment]
> I would like to do additional testing before this is eventually merged. Should I convert this pull request back to a draft, or is it better to add a Do Not Merge label?

either works

**Comment by WorldofJARcraft** [comment]
I found a small issue in the Ethernet driver (it allocated a buffer that was slightly too small).
I am confident that the DMA driver works.

**Review by teburd — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Pull request merged**
